File: server/src/messages/list.ts
import { map, without } from "lodash-es";

import Player from "@/players/player";

/**
 * A message that is sent when a player requests a list of players
 *
 * Sent on registration, on session end, and on request
 */
export interface ListMessage {
  type: "list";
  players: string[];
}

/**
 * List all players who aren't in a session (can be used to list all players. no reason to tho)
 *
 * Sent to a player on registration, on session end, and on request
 * @param {Player[]} players The players to list
 * @param {Player} requestedBy The player who requested the list event (their name won't be sent)
 * @returns {ListMessage} A List message
 */
function List(players: Player[], requestedBy: Player): ListMessage {
  const playersToSend = without(players, requestedBy);

  return {
    type: "list",
    players: map(playersToSend, "name"), // TODO: This is only for now. Some info should be sent
  };
}

export default List;

File: dotnet/src/Connection/List.cs
namespace Pong
{
  partial class Connection
  {
    /// <summary>
    /// A message that is sent when a player requests a list of players
    /// <br/>
    /// Sent on registration, on session end, and on request
    /// </summary>
    public class ListEvent : GameEvent
    {
      public List<string> Players { get; set; }

      public ListEvent(List<string> players)
        : base("list")
      {
        Players = players;
      }

      /// <summary>
      /// Deserialize a ListEvent from json
      /// </summary>
      public static new ListEvent Deserialize(string json)
      {
        return Deserialize<ListEvent>(json);
      }
    }

    /// <summary>
    /// Ask the server for a list of players
    /// </summary>
    public async Task List()
    {
      var listEvent = new Dictionary<string, string> { { "type", "list" } };

      await Send(listEvent);
    }
  }
}
